File: mac/Core/Protocol/KmFrame.swift
import Foundation

public enum KmType: UInt8, Equatable {
    case mouseMoveRel = 0x01
    case mouseMoveAbs = 0x02
    case mouseButton = 0x03
    case mouseScroll = 0x04
    case keyDown = 0x05
    case keyUp = 0x06
    case modifiersChanged = 0x07
    case heartbeat = 0xFE
    case clockSync = 0xFF
}

public struct KmFrame: Equatable {
    public let seq: UInt32
    public let timestampUs: UInt64
    public let type: KmType
    public let payload: Data

    public init(seq: UInt32, timestampUs: UInt64, type: KmType, payload: Data = Data()) {
        self.seq = seq
        self.timestampUs = timestampUs
        self.type = type
        self.payload = payload
    }
}

public enum KmFrameCodec {
    public static let headerSize = 13

    public static func encode(_ frame: KmFrame) -> Data {
        var data = Data(capacity: headerSize + frame.payload.count)
        data.appendLE(frame.seq)
        data.appendLE(frame.timestampUs)
        data.append(frame.type.rawValue)
        data.append(frame.payload)
        return data
    }

    public static func decode(_ data: Data) throws -> KmFrame {
        guard data.count >= headerSize else { throw BrpCodecError.truncated }
        let seq: UInt32 = data.readLE(at: 0)
        let ts: UInt64 = data.readLE(at: 4)
        let typeRaw = data.readU8(at: 12)
        guard let type = KmType(rawValue: typeRaw) else {
            throw BrpCodecError.unknownType(typeRaw)
        }
        let payloadStart = data.startIndex + headerSize
        let payload = Data(data[payloadStart..<data.endIndex])
        return KmFrame(seq: seq, timestampUs: ts, type: type, payload: payload)
    }
}

File: windows/Core/Protocol/KmFrame.cs
using System.Buffers.Binary;

namespace BarelyReal.Core.Protocol;

public enum KmType : byte
{
    MouseMoveRel = 0x01,
    MouseMoveAbs = 0x02,
    MouseButton = 0x03,
    MouseScroll = 0x04,
    KeyDown = 0x05,
    KeyUp = 0x06,
    ModifiersChanged = 0x07,
    Heartbeat = 0xFE,
    ClockSync = 0xFF
}

public sealed class KmFrame : IEquatable<KmFrame>
{
    public uint Seq { get; }
    public ulong TimestampUs { get; }
    public KmType Type { get; }
    public byte[] Payload { get; }

    public KmFrame(uint seq, ulong timestampUs, KmType type, byte[]? payload = null)
    {
        Seq = seq;
        TimestampUs = timestampUs;
        Type = type;
        Payload = payload ?? Array.Empty<byte>();
    }

    public bool Equals(KmFrame? other)
    {
        if (other is null) return false;
        return Seq == other.Seq
            && TimestampUs == other.TimestampUs
            && Type == other.Type
            && Payload.AsSpan().SequenceEqual(other.Payload.AsSpan());
    }

    public override bool Equals(object? obj) => obj is KmFrame other && Equals(other);
    public override int GetHashCode() => HashCode.Combine(Seq, TimestampUs, Type, Payload.Length);
}

public static class KmFrameCodec
{
    public const int HeaderSize = 13;

    public static byte[] Encode(KmFrame frame)
    {
        var buf = new byte[HeaderSize + frame.Payload.Length];
        var span = buf.AsSpan();
        BinaryPrimitives.WriteUInt32LittleEndian(span[..4], frame.Seq);
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(4, 8), frame.TimestampUs);
        span[12] = (byte)frame.Type;
        if (frame.Payload.Length > 0)
        {
            frame.Payload.AsSpan().CopyTo(span[HeaderSize..]);
        }
        return buf;
    }

    public static KmFrame Decode(ReadOnlySpan<byte> data)
    {
        if (data.Length < HeaderSize)
            throw new BrpCodecException(BrpCodecError.Truncated);

        var seq = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        var ts = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(4, 8));
        var typeRaw = data[12];

        if (!Enum.IsDefined(typeof(KmType), typeRaw))
            throw new BrpCodecException(BrpCodecError.UnknownType, typeRaw);

        var payload = data[HeaderSize..].ToArray();
        return new KmFrame(seq, ts, (KmType)typeRaw, payload);
    }
}
